src/utils/generator.js

```diff
@@ -52,12 +55,159 @@ import type {
   BabelNodeBlockStatement,
   BabelNodeLVal,
 } from "@babel/types";
-import { nullExpression, memberExpressionHelper } from "./babelhelpers.js";
-import { Utils, concretize } from "../singletons.js";
+import { memberExpressionHelper } from "./babelhelpers.js";
+import { concretize, Utils } from "../singletons.js";
 import type { SerializerOptions } from "../options.js";
+import * as t from "@babel/types";
+
+export type ResidualBuildNodeType =
+  | "IDENTIFIER"
+  | "SINGLE_ARG"
+  | "REBUILT_OBJECT"
+  | "CONSOLE_LOG"
+  | "FOR_IN"
+  | "DO_WHILE"
+  | "CONCRETE_MODEL"
+  | "BINARY_EXPRESSION"
+  | "LOGICAL_EXPRESSION"
+  | "CONDITIONAL_EXPRESSION"
+  | "UNARY_EXPRESSION"
+  | "ABSTRACT_FROM_TEMPLATE"
+  | "GLOBAL_ASSIGNMENT"
+  | "GLOBAL_DELETE"
+  | "EMIT_PROPERTY_ASSIGNMENT"
+  | "ABSTRACT_PROPERTY"
+  | "JOIN_GENERATORS"
+  | "APPEND_GENERATOR"
+  | "DEFINE_PROPERTY"
+  | "PROPERTY_DELETE"
+  | "THROW"
+  | "CONDITIONAL_THROW"
+  | "COERCE_TO_STRING"
+  | "ABSTRACT_FROM_TEMPLATE"
+  | "FOR_STATEMENT_FUNC"
+  | "NEW_EXPRESSION"
+  | "OBJECT_ASSIGN"
+  | "OBJECT_SET_PARTIAL"
+  | "OBJECT_GET_PARTIAL"
+  | "OBJECT_PROTO_HAS_OWN_PROPERTY"
+  | "OBJECT_PROTO_GET_OWN_PROPERTY_DESCRIPTOR"
+  | "ABSTRACT_OBJECT_SET_PARTIAL"
+  | "ABSTRACT_OBJECT_SET_PARTIAL_VALUE"
+  | "ABSTRACT_OBJECT_GET_PARTIAL"
+  | "ABSTRACT_OBJECT_GET_PROTO_OF"
+  | "ABSTRACT_OBJECT_GET"
+  | "DIRECT_CALL_WITH_ARG_LIST"
+  | "CALL_ABSTRACT_FUNC"
+  | "CALL_ABSTRACT_FUNC_THIS"
+  | "CALL_BAILOUT"
+  | "EMIT_CALL"
+  | "EMIT_CALL_AND_CAPTURE_RESULT"
+  | "GET_BINDING"
+  | "LOCAL_ASSIGNMENT"
+  | "LOGICAL_PROPERTY_ASSIGNMENT"
+  | "CONDITIONAL_PROPERTY_ASSIGNMENT"
+  | "PROPERTY_ASSIGNMENT"
+  | "MODULES_REQUIRE"
+  | "RESIDUAL_CALL"
+  | "ASSUME_CALL"
+  | "CANNOT_BECOME_OBJECT"
+  | "UPDATE_INCREMENTOR"
+  | "WIDENED_IDENTIFIER"
+  | "WIDEN_PROPERTY"
+  | "WIDEN_ABSTRACT_PROPERTY"
+  | "WIDEN_PROPERTY_ASSIGNMENT"
+  | "WIDEN_ABSTRACT_PROPERTY_ASSIGNMENT"
+  | "INVARIANT"
+  | "INVARIANT_APPEND"
+  | "DERIVED_ABSTRACT_INVARIANT"
+  | "PROPERTY_INVARIANT"
+  | "FULL_INVARIANT"
+  | "FULL_INVARIANT_FUNCTION"
+  | "FULL_INVARIANT_ABSTRACT"
+  | "UNKNOWN_ARRAY_METHOD_CALL"
+  | "UNKNOWN_ARRAY_METHOD_PROPERTY_CALL"
+  | "UNKNOWN_ARRAY_LENGTH"
+  | "UNKNOWN_ARRAY_GET_PARTIAL"
+  | "BABEL_HELPERS_OBJECT_WITHOUT_PROPERTIES"
+  | "REACT_DEFAULT_PROPS_HELPER"
+  | "REACT_TEMPORAL_FUNC"
+  | "REACT_CREATE_CONTEXT_PROVIDER"
+  | "REACT_SSR_REGEX_CONSTANT"
+  | "REACT_SSR_PREV_TEXT_NODE"
+  | "REACT_SSR_RENDER_VALUE_HELPER"
+  | "REACT_SSR_TEMPLATE_LITERAL"
+  | "REACT_NATIVE_STRING_LITERAL"
+  | "REACT_RELAY_MOCK_CONTAINER"
+  | "FB_MOCKS_BOOTLOADER_LOAD_MODULES"
+  | "FB_MOCKS_MAGIC_GLOBAL_FUNCTION";
+
+export type DerivedExpressionBuildNodeFunction = (
+  Array<BabelNodeExpression>,
+  SerializationContext,
+  Set<AbstractValue | ObjectValue>
+) => BabelNodeExpression;
+
+export type ResidualBuildNode = {
```

Comment:
The name `ResidualBuildNode` isn't very meaningful without historical knowledge.
I suggest something like `OperationDescriptor`.

Reply:
I'll rename that as part of this PR.